### DISCUSSION TITLE
NMM community edition integration

### DISCUSSION THREAD

**Discussion by Neshira** [focustense/easymod#45] [Ideas]
Hi, I was wondering if there was a chance for integration with NMM. 
The new community NMM has had virtual folder install for over a year now as well as compatibility with ESLs and literally every other feature that MO2 has (and vortex). It would be wonderful to have use of this tool for those of us who do not wish to use Vortex or MO2. Or perhaps a standalone version that doesn't rely on any mod managers and can be used for any setup. 

If not I completely understand, thank you anyway for making such a great tool.

**Comment by focustense** [comment]
Hi Neshira - and any others interested in the topic.

Thanks for the question and the interest. I'll leave any subjective debate over the merits of mod managers out of scope here, and stick to the technical aspects of what is required for EasyNPC to do its job effectively. If all of those things can be achieved with NMM Community, then maybe it already works with NMM Community. My guess is that it will not, but if people want to try anyway, I certainly won't get in the way.

Anyway: in order for an app like this to work, and to provide the kind of (I hope, and barring bugs) seamless user experience that it does, there are a few dependencies that I'll call table-stakes.

1. A file system for the game data that is _completely transparent_ to Windows APIs and the .NET Framework. Paths are paths are paths.
2. A mods directory, which has exactly one subdirectory per mod, and where each subdirectory contains all of the files for each mod in the same structure that was originally contained by the mod. Otherwise, like (1) above, I would have to start writing my own abstractions over the file system.
3. If the definition of "mod" in (2) is actually "mod file", there must be a reliable way to associate the "mod files" with their original mod names, and group together "mod file directories" that are actually part of the same mod. This is essential to both the mugshot resolution system, and generally being able to find the right assets and generate correct builds.
4. The ability to quickly and reliably disable and re-enable mods - not plugins but entire mods - and change the priorities of their files (not just their plugins). This is essential to conflict-avoidance and conflict-resolution in the EasyNPC world.
5. Not strictly required, but a strong profile system is ideal for eliminating a lot of the tedium of (4) above.

Don't take this as an exhaustive list - it's a baseline.

For the above points:

1. Mod Organizer uses a vfs driver and Vortex uses hardlinks.
2. Mod Organizer has a mods directory and Vortex has a mod staging directory.
3. Mod Organizer supports (and recommends) merging mod files into one mod. Vortex maintains metadata from the Nexus and has an extension system that enables passing that info to EasyNPC.
4. This is a single instant click in Mod Organizer; Vortex needs a separate deployment step but it's usually automatic.
5. Mod Organizer and Vortex both have a profile system (Vortex, I'm told, is a little harder to use.)

NMM Community may have changed a great deal in recent years. For point (1), I believe it uses symlinks now, which is far better for overall game stability and mod maintenance than copying into the data directory, but I am not sure if it is transparent to Windows APIs. That's not to say that it _isn't_ transparent, but I don't know and am not prepared to claim official support or take on bug-fixing if it doesn't; but if it works out of the box, then great!

Regarding the other points, NMM was very weak on all of them several years ago, but again, maybe those things have changed. Feel free to try it out and see. EasyNPC doesn't try to lock you out if you're using an unsupported mod manager, I just can't promise it'll work correctly. Two good proxy tests would be Synthesis and zMerge - if you're able to run both these tools on an NMM-managed game without any issues, there's a good chance EasyNPC will work as well.

In theory, you don't need any mod manager to use EasyNPC. You can manually unpack your mods to an arbitrary directory on your computer, and manually deploy all your files to the Skyrim data directory, and as long as you point EasyNPC to the former, and keep the latter properly maintained, it'll do fine. I don't think it would be a _good user experience_, and don't want to promote and stand behind what I'd consider to be a bad user experience, which is why I'm not listing this as an option or providing instructions for it. It can already be a challenge getting things to work _with_ a sophisticated mod manager.

So really there are 3 questions to separate here:

- **Will EasyNPC work with NMM Community as it is currently designed/implemented?** - Maybe.
- **If not, could it be modified to support NMM Community?** - Maybe. It's open source, and I have no objections to people trying - not that it would matter even if I did object, because GPL. Whether I'd accept a pull request depends on the overall complexity, potential risk for MO2/Vortex users, and potential for longer-term support issues.
- **Will I add it to the official support list and personally maintain that support?** - No, supporting Vortex along with MO2 is already a challenge and NMM isn't on the roadmap. Even if NMM meets every requirement with flying colors, it's still too much work at the moment to support a 3rd mod manager. If other people are contributing, that might change.

It is probably not the answer you were looking for, but I hope this makes the reasoning clear. Again, I appreciate the question and the interest!

**Comment by Neshira** [comment]
Thank you so much for your thorough reply! I completely understand that three mod managers is too many to support. NMM has changed SIGNIFICANTLY in the past few years, especially this year. While the overall user interface is similar to the past it's practically a different program under the hood and it can utilize zmerge and synthesis so it might work with EasyNPC! I'll give it a shot and let you know. Even if it's unsupported at least then you'll know if it's possible!
